### PR TITLE
[BugFix] Finish async_chunk requests without pad-token injection

### LINKED ISSUE #3607: [Bug]: "Code2Wav input_ids length 1" warning emitted for every request on async_chunk Qwen3-TTS, regardless of caller
## Summary

vllm-omni 0.20.0 emits the following `WARNING` **on every request** when serving Qwen3-TTS in `async_chunk: true` mode:

```
WARNING [qwen3_tts_code2wav.py:156] Code2Wav input_ids length 1 not divisible by num_quantizers 16; skipping malformed request.
```

The warning is **benign** -- the request still finishes and previously-decoded audio chunks are returned correctly -- but it produces high log volume in production and the "malformed request" wording falsely implies a client error.

## Reproduction

Any caller of `AsyncOmni.generate()` on a Qwen3-TTS-12Hz model with `async_chunk: true` reproduces it. Verified across:

- `vllm-omni serve --omni Qwen/Qwen3-TTS-12Hz-1.7B-CustomVoice` (CLI subprocess hitting `/v1/audio/speech`)
- `OmniOpenAIServingSpeech.create_speech` reused in-process
- Direct `AsyncOmni.generate(...)` from a plain Python script

The warning fires per request in all three. None of the following caller-side variations changed that:

- `tokens_input([1]*ph_len)` vs plain dict prompt construction
- `coerce_param_message_types(list(defaults), is_streaming=True)` vs handrolled `output_kind=DELTA`
- `stop_token_ids: [2150]` on stage-0 default sampling params
- Adding `voice_created_at: [0]` to `additional_information`

## Why the warning exists (what it actually validates)

`num_quantizers=16` is the codebook count of the Qwen3-TTS-12Hz residual vector quantizer (RVQ): each codec frame is encoded as 16 parallel codebook indices, one per quantizer level. Code2Wav's `forward()` receives a flat 1-D tensor `input_ids` whose length must be `num_frames x num_quantizers` so the codes can be reshaped into `[num_frames, 16]` and fed to the decoder. The divisibility check at `qwen3_tts_code2wav.py:156` enforces this invariant -- a length not divisible by 16 means the input doesn't represent a whole number of frames, so the decoder cannot produce a coherent waveform and the request is skipped (an empty audio buffer is appended). The warning is correct **as a defensive guard**: a length-1 input is genuinely malformed for code2wav. The bug is upstream of the guard -- the pipeline is producing a length-1 input on every request via the EOS pad-zero injection (traced below), which the guard then catches.

## Root cause trace

End-of-stream pad-zero injection at the talker -> code2wav handoff. File:line at v0.20.0:

1. **Talker finishes between codec-chunk boundaries.** `_extract_last_frame` (`stage_input_processors/qwen3_tts.py:127-128`) rejects the all-zero EOS frame; no frame is appended.

2. **Empty-codes-finish payload.** `talker2code2wav_async_chunk` (`stage_input_processors/qwen3_tts.py:209-214`) sees `length <= 0 and finished` and emits:
   ```python
   {"codes": {"audio": []}, "meta": {"finished": True}}
   ```

3. **Receiver writes empty `prompt_token_ids`.** `chunk_transfer_adapter.py:165-166` sets `request.prompt_token_ids = []` and adds the request to `finished_requests`.

4. **Scheduler pad-zero injection.** `OmniGenerationScheduler.schedule()` (`core/sched/omni_generation_scheduler.py:95-105`, `:150-156`) sees `len(prompt_token_ids) == 0` for a finished request and injects a single `0` so the stage can emit a FINISHED `EngineCoreOutput`.

5. **Code2Wav rejects the 1-token input.** `Qwen3TTSCode2Wav.forward()` (`model_executor/models/qwen3_tts/qwen3_tts_code2wav.py:147-162`) sees `input_ids = tensor([0])`, fails `n % num_quantizers == 0` (1 is not a multiple of 16), logs the warning, sets `audios[idx] = torch.zeros(0)`, continues.

Audio output is correct because earlier valid chunks have already been decoded; only this terminal forward pass is wasted.

## Related

- #3124 -- exact warning title, closed `invalid`. Different surface cause (Base voice-clone with mismatched ref-text).
- #2911 -- moved `RequestOutputKind.DELTA` enforcement out of `AsyncOmni.generate` into the entrypoint. Required by direct `AsyncOmni.generate` callers; unrelated to this warning.
- #3396 -- analogous Qwen2.5-Omni fix defensively skipping empty codec input updates. Same shape of fix, different model.

## Environment

- vllm-omni: 0.20.0 (PyPI)
- vllm: 0.20.2
- Model: `Qwen/Qwen3-TTS-12Hz-1.7B-CustomVoice` (also reproduced on `0.6B-CustomVoice`)
- Deploy YAML: bundled `vllm_omni/deploy/qwen3_tts.yaml`
- GPU: NVIDIA L40S (sm_89)

### PR DESCRIPTION
**Pull request by NickCao**
## Purpose

When an async_chunk upstream stage completes with no remaining codec tokens, the scheduler previously injected a single pad [0] into prompt_token_ids to force one more forward pass so the FINISHED condition could trigger. This caused Code2Wav to warn about a length-1 input not being divisible by num_quantizers on every request.

Instead, collect these requests in _pending_finish_reqs during schedule() and emit their FINISHED EngineCoreOutput directly in update_from_output(), bypassing the model forward pass entirely.

Fixes https://github.com/vllm-project/vllm-omni/issues/3607

## Test Plan

  - pytest tests/distributed/omni_connectors/test_chunk_transfer_adapter.py -v — 25 passed, 2 failed
  (pre-existing on main)
  - pytest tests/e2e/offline_inference/test_qwen3_tts_base.py -v — 1 passed
  - pytest tests/e2e/offline_inference/test_qwen3_tts_customvoice.py -v — 1 passed
  - pytest tests/e2e/offline_inference/test_cosyvoice3_expansion.py -v — skipped (main also fails)
  - pytest tests/e2e/online_serving/test_qwen3_tts_base.py -v — 2 passed
  - pytest tests/e2e/online_serving/test_qwen3_tts_base_expansion.py -v — 4 passed
  - pytest tests/e2e/online_serving/test_qwen3_tts_customvoice.py -v — 2 passed
  - pytest tests/e2e/online_serving/test_qwen3_tts_customvoice_expansion.py -v — 8 passed
  - pytest tests/e2e/online_serving/test_qwen3_omni.py -v — 3 passed
  - pytest tests/e2e/online_serving/test_qwen3_omni_expansion.py -v — 35 passed, 3 skipped;
  test_text_video_to_text_audio_001[async_chunk] deselected (hangs on main)
  - pytest tests/e2e/online_serving/test_cosyvoice3_tts_expansion.py -v — 3 passed
  - pytest tests/e2e/accuracy/qwen3_omni/test_qwen3_omni.py -v — not run (WER bench threshold,
  pre-existing failure)

## Test Result

Done, no regressions.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/core/sched/omni_generation_scheduler.py [hunk 1]
```diff
@@ -520,6 +510,33 @@ def update_from_output(
...
+                    EngineCoreOutput(
+                        request_id=request.request_id,
+                        new_token_ids=[],
+                        finish_reason=request.get_finished_reason(),
+                        stop_reason=request.stop_reason,
+                        events=request.take_events(),
+                        trace_headers=request.trace_headers,
+                    )
+                )
+            stopped_running_reqs.add(request)
```
Comment (linyueqian) [inline vllm_omni/core/sched/omni_generation_scheduler.py:601]: [important] The `outputs[...].append(...)` is gated on `finished`. When `_handle_stopped_request` returns `False` the request still gets:

* `status = FINISHED_STOPPED`,
* added to `stopped_running_reqs` (so it leaves `self.running`),

but no `EngineCoreOutput` is emitted and `chunk_transfer_adapter.cleanup` is skipped. The engine has dropped the request server-side; the client never receives a finish token and the stream hangs until its own timeout fires.

The existing stopped-request path at lines 491-508 emits the output for any `stopped` request regardless of `finished` (with `finish_reason` already populated). That pattern keeps the engine and client view consistent.

For async_chunk requests that bypass the forward pass there is usually no KV xfer pending, so `_handle_stopped_request` should return `True` in practice today. But that is a load-bearing assumption rather than a guarantee. If a future config attaches a KV or EC connector to async_chunk requests we silently lose finish notifications with no error log.

Suggest hoisting `outputs[request.client_index].append(...)` out of the `if finished:` block so it always fires once `request.status == FINISHED_STOPPED`, matching the existing path.

Minor on the same block: `_free_request(request)` returns `kv_transfer_params` and the main path forwards it into the output; here the return value is discarded. Fine for async_chunk today, inconsistent with the contract of the main path.

Comment (NickCao) [reply vllm_omni/core/sched/omni_generation_scheduler.py:601]: Fixed.
